src/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -85,7 +81,7 @@ public static CallSite Create(Type delegateType, CallSiteBinder binder)
             ContractUtils.RequiresNotNull(binder, nameof(binder));
             if (!delegateType.IsSubclassOf(typeof(MulticastDelegate))) throw System.Linq.Expressions.Error.TypeMustBeDerivedFromSystemDelegate();
 
-            var ctors = s_siteCtors;
+            CacheDict<Type, Func<CallSiteBinder, CallSite>> ctors = s_siteCtors;
             if (ctors == null) {
                 // It's okay to just set this, worst case we're just throwing away some data
                 s_siteCtors = ctors = new CacheDict<Type, Func<CallSiteBinder, CallSite>>(100);
```

Comment:
Incidentally, if this approach to races is taken, is it worth s_siteCtors being volatile? Or maybe not being volatile but CASing on the write (which IIRC, fences) would in fact be lighter in the more common already-set case?